### PR TITLE
Adding Read support for json & jsonb datatypes in postgresql

### LINKED ISSUE #6614: Column '"data"' cannot be resolved for jsonb
created a jsonb table in postgres db and trying to query it from presto and its showing below error.

**in postgresdb:**
CREATE TABLE json_test (
  id serial primary key,
  data jsonb
);

INSERT INTO json_test (data) VALUES 
  ('{}'),
  ('{"a": 1}'),
  ('{"a": 2, "b": ["c", "d"]}'),
  ('{"a": 1, "b": {"c": "d", "e": true}}'),
  ('{"b": 2}');

SELECT * FROM json_test;

**in presto** 

presto:public> SELECT * FROM json_test;
 id
----
  1
  2
  3
  4
  5
(5 rows)

Query 20161114_162548_00003_9fxak, FINISHED, 2 nodes
Splits: 2 total, 2 done (100.00%)
0:00 [5 rows, 0B] [53 rows/s, 0B/s]

presto:public> SELECT * FROM json_test WHERE data = '{"a":1}';
Query 20161114_162602_00004_9fxak failed: line 1:31: Column '"data"' cannot be resolved
SELECT * FROM json_test WHERE data = '{"a":1}'

 presto:public> SELECT * FROM json_test WHERE data = '{"a":1}';
Query 20161114_162933_00006_9fxak failed: line 1:31: Column '"data"' cannot be resolved
SELECT * FROM json_test WHERE data = '{"a":1}'

### PR DESCRIPTION
**Pull request by agrawalreetika**
Adding read support for json & jsonb datatypes in postgresql
Fixes #6614

### CODE REVIEW

**Comment by aweisberg** [comment]
Thanks for contributing this. Please take a look at our attribution guidelines in case any of this code is coming from PrestoSQL/TrinoDB https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution

I can definitely see a lot of work went into getting this adapted to PrestoDB, but we try to go above and beyond at crediting people.

**Comment by agrawalreetika** [comment]
> Thanks for contributing this. Please take a look at our attribution guidelines in case any of this code is coming from PrestoSQL/TrinoDB https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution
> 
> I can definitely see a lot of work went into getting this adapted to PrestoDB, but we try to go above and beyond at crediting people.

**Comment by aweisberg** [comment]
OK, thanks! Sorry I missed you had already done it.

**Review by rongrong — COMMENTED** [review]
Can we add some tests for the new code in `PostgreSqlClient`? The commit in trino does include some tests.

**Comment by agrawalreetika** [comment]
> Can we add some tests for the new code in `PostgreSqlClient`? The commit in trino does include some tests.

@rongrong  Thank you for your review. Tests are added.

**Comment by rongrong** [comment]
If you click on the travis CI build link, you'll see the newly added tests failed:
```
[ERROR] Failures: 
[ERROR]   TestPostgreSqlTypeMapping.jsonDataTypeTest:288 » Runtime Error executing sql:
...
[ERROR]   TestPostgreSqlTypeMapping.testJson:301->AbstractTestQueryFramework.assertUpdate:211->AbstractTestQueryFramework.assertUpdate:216 » Runtime
```
You should be able to run these tests in InteliJ. So please make sure the tests pass and update the change. Thanks for working on this!

File: presto-postgresql/src/main/java/com/facebook/presto/plugin/postgresql/PostgreSqlClient.java [hunk 1]
```diff
@@ -83,6 +113,20 @@ protected String toSqlType(Type type)
         return super.toSqlType(type);
     }
 
+    @Override
+    public Optional<ReadMapping> toPrestoType(ConnectorSession session, JdbcTypeHandle typeHandle)
+    {
+        switch (typeHandle.getJdbcType()) {
```
Comment (rongrong) [inline presto-postgresql/src/main/java/com/facebook/presto/plugin/postgresql/PostgreSqlClient.java:119]: I'm not sure this operation is safe. `CLOB` can be any text and doesn't need to be json right?

File: presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlTypeMapping.java [hunk 2]
```diff
@@ -267,13 +270,65 @@ public void testTimestamp()
...
+                .addRoundTrip(jsonDataType, "{\"a\":[1,2,3],\"b\":{\"aa\":11,\"bb\":[{\"a\":1,\"b\":2},{\"a\":0}]}}")
+                .addRoundTrip(jsonDataType, "[]");
+    }
+
+    private static DataType<String> jsonDataType()
+    {
+        return dataType(
+                "json",
+                JSON,
+                value -> "JSON " + formatStringLiteral(value),
```
Comment (rongrong) [inline presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlTypeMapping.java:307]: The data format of `json` and `jsonb` could be different right? Can we show that and make sure we test the differences as well?

File: presto-postgresql/src/main/java/com/facebook/presto/plugin/postgresql/PostgreSqlClient.java (hunk 1 above)
Comment (agrawalreetika) [reply presto-postgresql/src/main/java/com/facebook/presto/plugin/postgresql/PostgreSqlClient.java]: Yeah, I CLOB changes are not really required in here. I have taken that out

File: presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlTypeMapping.java (hunk 2 above)
Comment (agrawalreetika) [reply presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlTypeMapping.java:307]: I had added test cases for both json and jsonb data format. If you think we have to add something more, could you please give me the reference?

Comment (rongrong) [reply presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlTypeMapping.java:307]: I don't know all the details about the difference between json an jsonb. But i think one of the difference is extra spaces. What I mean here is that the data provided for both json and jsonb are the same. Can we format them so the test data for json are not valid jsonb data, and vice versa?

Comment (agrawalreetika) [reply presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlTypeMapping.java:307]: @rongrong Sorry for the delay in this. Currently, Postgres JSON and JSONB data types are mapped to Presto JSON type.

**Comment by rohanpednekar** [comment]
Hi @rongrong could you please do a final pass on this?

**Comment by rohanpednekar** [comment]
Hi @rongrong, could you please do a final pass on this? We have multiple community users asking for this. It would be great if we can add it to the next release.

**Comment by rohanpednekar** [comment]
Gentle Ping! 🙏

**Comment by d9k** [comment]
@rongrong, pretty convincing ping

**Comment by d9k** [comment]
@agrawalreetika, I'm not good at java deployment. Is there a way to add your code patches into the current `presto-server-${PRESTO_VERSION}.tar.gz` and `presto-cli-${PRESTO_VERSION}-executable.jar`?

**Comment by agrawalreetika** [comment]
> @agrawalreetika, I'm not good at java deployment. Is there a way to add your code patches into the current `presto-server-${PRESTO_VERSION}.tar.gz` and `presto-cli-${PRESTO_VERSION}-executable.jar`?

@d9k We can chat over slack. I can help you.

**Comment by rohanpednekar** [comment]
@tdcmeehan Could you please help us to get the final review and get this merged? Thanks much!

**Comment by rongrong** [comment]
Sorry not sure how I missed this before. Can we rebase to the latest? Also one of my earlier comment was not addressed. There were 2 unit tests added in the original Trino PR, only one is included here. Please add the other one (with the queries) as well. Thanks!

**Comment by agrawalreetika** [comment]
Hi @rongrong, Thank you for your review on PR!
- I have rebased the PR
- Regarding other test case for `prestoCreateAsSelect` Presto JSON column needs to set to PGobject for Postgres connector. I saw the current JdbcPageSink has that. In trino, the writing column can be extended to the respective JDBC connector. Once we add generic column mapping in prestodb as well we can extend the write su[p[ort for Json column in Postgres connector.

**Comment by rongrong** [comment]
> Hi @rongrong, Thank you for your review on PR!
> 
> * I have rebased the PR
> * Regarding other test case for `prestoCreateAsSelect` Presto JSON column needs to set to PGobject for Postgres connector. I saw the current JdbcPageSink has that. In trino, the writing column can be extended to the respective JDBC connector. Once we add generic column mapping in prestodb as well we can extend the write su[p[ort for Json column in Postgres connector.

So this PR only supports read? Can we be specific in the PR and commit title to reflect that then? Thanks!

**Comment by agrawalreetika** [comment]
Yes, that's right it will support only Read. I will update the Commit and PR Title

**Comment by rohanpednekar** [comment]
Thanks @agrawalreetika and @rongrong!

**Pull request merged**